tikv/raftstore/peer.go

```diff
@@ -992,3 +1028,930 @@ func (p *Peer) sendRaftMessage(msg eraftpb.Message, trans chan<- *rspb.RaftMessa
+			p.PostPropose(meta, false, EmptyCallback)
+		}
+	}
+
+	return true
+}
+
+func (p *Peer) GetMinProgress() uint64 {
+	var minMatch uint64 = math.MaxUint64
+	hasProgress := false
+	for _, pr := range p.RaftGroup.Status().Progress {
+		hasProgress = true
+		if pr.Match < minMatch {
+			minMatch = pr.Match
+		}
+	}
+	if !hasProgress {
+		return 0
+	}
+	return minMatch
+}
+
+func (p *Peer) preProposePrepareMerge(ctx *PollContext, req *raft_cmdpb.RaftCmdRequest) error {
+	lastIndex := p.RaftGroup.Raft.RaftLog.LastIndex()
+	minProgress := p.GetMinProgress()
+	minIndex := minProgress + 1
+	if minProgress == 0 || lastIndex-minProgress > ctx.Cfg.MergeMaxLogGap {
+		return fmt.Errorf("log gap (%v, %v] is too large, skip merge", minProgress, lastIndex)
+	}
+
+	entrySize := 0
+
+	ents, err := p.RaftGroup.Raft.RaftLog.Entries(minIndex, math.MaxUint64)
+	if err != nil {
+		return err
+	}
+	for _, entry := range ents {
+		entrySize += len(entry.Data)
+		if entry.EntryType == eraftpb.EntryType_EntryConfChange {
+			return fmt.Errorf("log gap contains conf change, skip merging.")
+		}
+		if len(entry.Data) == 0 {
+			continue
+		}
+		cmd := raft_cmdpb.RaftCmdRequest{}
+		err := cmd.Unmarshal(entry.Data)
+		if err != nil {
+			panic(fmt.Sprintf("%v data is corrupted at %v, error: %v", p.Tag, entry.Index, err))
+		}
+		if cmd.AdminRequest == nil {
+			continue
+		}
+		cmdType := cmd.AdminRequest.GetCmdType()
+		switch cmdType {
+		case raft_cmdpb.AdminCmdType_TransferLeader, raft_cmdpb.AdminCmdType_ComputeHash,
+			raft_cmdpb.AdminCmdType_VerifyHash, raft_cmdpb.AdminCmdType_InvalidAdmin:
+			continue
+		default:
+		}
+
+		// Any command that can change epoch or log gap should be rejected.
+		return fmt.Errorf("log gap contains admin request %v, skip merging.", cmdType)
+	}
+
+	if float64(entrySize) > float64(ctx.Cfg.RaftEntryMaxSize)*0.9 {
+		return fmt.Errorf("log gap size exceed entry size limit, skip merging.")
+	}
+
+	req.AdminRequest.PrepareMerge.MinIndex = minIndex
+	return nil
+}
+
+func (p *Peer) PrePropose(pollCtx *PollContext, req *raft_cmdpb.RaftCmdRequest) (*ProposalContext, error) {
+	if err := pollCtx.CoprocessorHost.PrePropose(p.Region(), req); err != nil {
+		return nil, err
+	}
+	ctx := new(ProposalContext)
+
+	if getSyncLogFromRequest(req) {
+		ctx.insert(ProposalContext_SyncLog)
+	}
+
+	if req.AdminRequest == nil {
+		return ctx, nil
+	}
+
+	switch req.AdminRequest.GetCmdType() {
+	case raft_cmdpb.AdminCmdType_Split, raft_cmdpb.AdminCmdType_BatchSplit:
+		ctx.insert(ProposalContext_Split)
+	default:
+	}
+
+	if req.AdminRequest.PrepareMerge != nil {
+		err := p.preProposePrepareMerge(pollCtx, req)
+		if err != nil {
+			return nil, err
+		}
+		ctx.insert(ProposalContext_PrepareMerge)
+	}
+
+	return ctx, nil
+}
+
+func (p *Peer) ProposeNormal(pollCtx *PollContext, req *raft_cmdpb.RaftCmdRequest) (uint64, error) {
+	if p.PendingMergeState != nil && (req.AdminRequest == nil || req.AdminRequest.CmdType != raft_cmdpb.AdminCmdType_RollbackMerge) {
```

Comment:
We can call `req.AdminRequest.GetCmdType()` directly without checking if `AdminRequest` is nil.

Reply:
No, if `AdminRequest` is nil, here will panic, but TiKV will not panic.